draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1190,213 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+                    is_leader: bool,
+                    agg_param: Vdaf.AggParam,
+                    nonce: bytes[Vdaf.NONCE_SIZE],
+                    public_share: bytes,
+                    host_input_share: bytes) -> State:
+    try:
+        prep_state = Vdaf.prep_init(
+            vdaf_verify_key,
+            0 if is_leader else 1,
+            agg_param,
+            nonce,
+            public_share,
+            host_input_share,
+        )
+        return Continue(0, prep_state)
+    except:
+        return Rejected()
+~~~
+
+If the state is `Rejected`, then processing halts. Otherwise, if the state is
+`Continued`, then processing continues using the following transition rule until
+a terminal state is reached.
+
+Let us call the initiating party the "Leader" and the responding party the
+"Helper". The high-level idea is that the Leader and Helper will take turns
+running the computation locally until input from their peer is required:
+
+* For a 1-round VDAF (e.g., Prio3 ({{prio3}})), the Leader sends its prep share
+  to the Helper, who computes the prep message locally, computes its output
+  share, then sends the prep message to the Leader. Preparation requires just
+  one round trip between the Leader and the Helper.
+
+* For a 2-round VDAF (e.g., Poplar1 ({{poplar1}})), the Leader sends its
+  first-round prep share to the Helper, who replies with the first-round prep
+  message and its second-round prep share. In the next request, the Leader
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two round trips.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+the last round is reached and each Aggregator computes an output share. All
+told there there are `ceil((Vdaf.ROUNDS+1)/2)` requests sent.
+
+Each Aggregator handles these messages according to the following transition
+rule, so long as they are in the `Continued` state:
+
+~~~ transition
+def ping_pong_next(Vdaf,
+                   is_leader: bool,
+                   agg_param: Vdaf.AggParam,
+                   state: State,
+                   inbound: Optional[bytes],
+                   ) -> (State, Optional[PingPongMessage]):
+    round, host_prep_state = state.round, state.prep_state
+    if inbound != None:
+        if round == 0:
+            if is_leader:
+                return (Rejected(), None)
+            else:
+                prep_msg, peer_prep_share = None, inbound
+        elif round == Vdaf.ROUNDS:
```

Comment:
I'll think on it more, but I'm not totally convinced we can do away with the round number altogether. In any case I think we can make parsing context-free here.

A couple of points:
* I'm not sure notwork failures are or should be a consideration for VDAF. Ideally we can think of VDAF as the primitive building block that is used in a protocol that deals with this stuff (DAP). As an anology, we don't typically think about network failures when writing down the spec for Diffie-Hellman key exchange. (And I'd argue we don't need to.)
* Elsewhere in VDAF, parsing is not context-free: For instance in Prio3 the field being used is not encoded in any message. I'm fine doing so here however, since this is connective tissue for a specific protocol.

Reply:
i've added the discriminator byte and removed the round number from the Continue state once more.